### DISCUSSION TITLE
Representation of Gates/Pulses

### DISCUSSION THREAD

**Discussion by pbethke** [qutech/qupulse#727] [General]
Gates are logical operations on a single or multiple qubits. Their associated control signals are called pulses. The two concepts are strongly connected. Gates and pulses have the following properties:
## Gates
- Gates are unitary operations performed on one or multiple qubits
- Gates can be concatenated to produce new gates
- Qubits usually have a set of gates that they can perform, forming a complete basis from which to compose any gate.
- The gates in a physical qubit's basis gate set each have an associated _pulse_
## Pulses
- Pulses represent the actual control signal on the qubit
- Pulses map onto commands for the physical qubit control electronics
## Problem

pulse and gate definitions somewhat overlap. We should find one term for it and stick to it. The way I try to separate the two is to always use _gate_ down to the physical qubit level, where _pulse_ becomes appropriate. _Gates_ are then logical operations on qubits and may be static, whereas _pulses_ depend on the qubit it is acting on, calibration, etc.
## Qubit Hierarchy and Gates/Pulses

The top level corresponds to _algorithms_, the middle one to things like simulated versions of physical qubits or logical qubits and the bottom level is the level of real physical qubits, calibration and pulses.
![g](https://cloud.githubusercontent.com/assets/10436440/6689832/17533cd4-ccbd-11e4-9935-1e6cec5a9ab8.png)

**Comment by hbluhm** [comment]
It seems worth discussing if gates/algorithms (which also include other operations) and pulses/waveforms should be combined into one type of object or hierarchy or be separate entities representing the mathematical model on the one hand and control pulse on the other.

Also, I would suggest to only introduce different classes for different hierarchy levels if a natural, sharp distinction exists. (I am currently not aware of one.) Otherwise, operations and waveforms could be generic designations.

**Comment by pbethke** [comment]
the only real distinction here is that the gates that comprise the gate set of a physical qubit should map onto corresponding pulses and are not mere mathematical objects. One could argue of course, that a control signal itself is also a gate, but its form depends on calibration of the actual qubit, so its not as _static_ as other gates.
